### PR TITLE
feat(sdk): integrate with scroll-proving-sdk

### PR DESCRIPTION
**Pull request by ErdongChen61**
Integrate with [scroll-proving-sdk](https://github.com/scroll-tech/scroll-proving-sdk) to support proof generation for Scroll SDK users.

The high level idea is that this binary will fetch proving tasks from Scroll Coordinator and then send requests to the Snarkify platform for proof generation. And it will check the job status periodically and submit the results to Scroll Coordinator at the end.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+            .map_err(serde::de::Error::custom)
+    })
+    .transpose()
+}
+
+#[derive(Parser, Debug)]
+#[clap(disable_version_flag = true)]
+struct Args {
+    /// Path to the configuration file in JSON format.
+    /// Please refer to config.json in https://github.com/snarkify/scroll-proving-sdk
```
Comment (cyphersnake) [inline src/main.rs]: I don't see anything relevant in the link

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+    #[arg(long = "config", default_value = "config.json")]
+    config_file: String,
+    /// Unique UUID for the service in Snarkify platform.
+    #[arg(long = "service-id")]
+    service_id: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct SnarkifyGetTaskResponse {
+    /// Task UUID in Snarkify platform.
```
Comment (cyphersnake) [inline src/main.rs]: Using an empty string is not good practice, however, if you want to keep it, please indicate in the rustdoc that there can be either an empty string or uuid here

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+    pub started: Option<DateTime<Utc>>,
+    #[serde(deserialize_with = "deserialize_datetime")]
+    pub finished: Option<DateTime<Utc>>,
+    pub state: SnarkifyTaskState,
+    /// Task input data necessary for the proof generation.
+    pub input: String,
+    /// Task proof.
+    pub proof: Option<String>,
+    /// Task error message.
+    pub error: Option<String>,
```
Comment (cyphersnake) [inline src/main.rs]: Such rustdoc are not useful and can be removed. 

However, normal rustdocs can be very useful. For example, "serialized in hex binary proof" could simplify parsing of this code by an order of magnitude

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+    pub proof_type: Option<SnarkifyProofType>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct SnarkifyGetVkResponse {
+    /// Verifying key.
```
Comment (cyphersnake) [inline src/main.rs]: Similarly, I know what `vk` looks like in different proof-systems, but I don't understand what format they are in here. Serialized? Encoded?

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+    fn build_url(&self, method: &str) -> anyhow::Result<Url> {
+        let full_url = format!("{}{}", self.base_url, method);
+        Url::parse(&full_url).map_err(|e| anyhow::anyhow!(e))
+    }
+
+    async fn get_with_token<Resp>(&self, method: &str) -> anyhow::Result<Resp>
+    where
+        Resp: serde::de::DeserializeOwned,
+    {
+        let url = self.build_url(method)?;
```
Comment (cyphersnake) [inline src/main.rs]: If you get a parsing error here, you lose all the context when throwing the error up. 

You need to either fill out the context (inside anyhow macro) or add a log

**Review by cyphersnake — CHANGES_REQUESTED** [review]
In general, my comments are about the same thing. Using strings in web applications is a necessary evil, however, the main disadvantage of this approach - unpredictability of the content, can be mitigated with the help of competent rustdocs

These rustdoc should not duplicate the field name but report the contents as much. as we know.

File: src/main.rs (hunk 1 above)
Comment (ErdongChen61) [reply src/main.rs]: good catch! Updated the comments

File: src/main.rs (hunk 2 above)
Comment (ErdongChen61) [reply src/main.rs]: sg!

File: src/main.rs (hunk 4 above)
Comment (ErdongChen61) [reply src/main.rs]: sg!

File: src/main.rs (hunk 3 above)
Comment (ErdongChen61) [reply src/main.rs]: sg!

File: src/main.rs (hunk 5 above)
Comment (ErdongChen61) [reply src/main.rs]: good catch!

**Review by cyphersnake — CHANGES_REQUESTED** [review]
WIP

File: src/datetime_utils.rs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+{
+    // The datetimes from the Snarkify API does not provide timezone information,
+    // so we assume it is UTC.
+    let s: Option<String> = Option::deserialize(deserializer)?;
+    s.map(|s| {
+        NaiveDateTime::parse_from_str(&s, "%Y-%m-%dT%H:%M:%S")
+            .map(|dt| DateTime::<Utc>::from_naive_utc_and_offset(dt, Utc))
+            .map_err(serde::de::Error::custom)
+    })
+    .transpose()
```
Comment (cyphersnake) [inline src/datetime_utils.rs]: NBD:
```rust
Option::<String>::deserialize(deserializer)?
    .map(|s| {
        NaiveDateTime::parse_from_str(&s, "%Y-%m-%dT%H:%M:%S")
            .map(|dt| DateTime::<Utc>::from_naive_utc_and_offset(dt, Utc))
            .map_err(serde::de::Error::custom)
    })
    .transpose()
```

File: src/prover.rs [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+};
+
+use crate::types::{
+    SnarkifyCreateTaskInput, SnarkifyCreateTaskRequest, SnarkifyGetTaskResponse,
+    SnarkifyGetVkResponse,
+};
+use serde::Serialize;
+
+/// API version used by the Snarkify platform.
+const API_VERSION: &'static str = "v1";
```
Comment (cyphersnake) [inline src/prover.rs]: NBD

`'static` not needed here
```rust
const API_VERSION: &str = "v1";
```

File: src/types.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+use chrono::{DateTime, Utc};
+use scroll_proving_sdk::prover::proving_service::ProveRequest;
+use scroll_proving_sdk::prover::types::CircuitType;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Debug)]
+pub struct SnarkifyGetVkResponse {
+    /// Base64 encoded verification key.
+    pub vk: String,
+}
```
Comment (cyphersnake) [inline src/types.rs]: I really don't want to torture you with this again, but it's still unclear what might lie here! I.e. encoding is great (really), but what is the format of the encoded data inside? If it's arbitrary, what does it depend on? etc

**Review by cyphersnake — CHANGES_REQUESTED** [review]

File: src/prover.rs (hunk 7 above)
Comment (ErdongChen61) [reply src/prover.rs]: sg!

File: src/datetime_utils.rs (hunk 6 above)
Comment (ErdongChen61) [reply src/datetime_utils.rs]: sg!

File: src/types.rs (hunk 8 above)
Comment (ErdongChen61) [reply src/types.rs]: added the usage of the vk.

File: src/prover.rs [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    send_timeout: Duration,
+    client: ClientWithMiddleware,
+}
+
+#[async_trait]
+impl ProvingService for SnarkifyProver {
+    fn is_local(&self) -> bool {
+        false
+    }
+    async fn get_vk(&self, req: GetVkRequest) -> GetVkResponse {
```
Comment (Gigatron) [inline src/prover.rs:41]: missing  new line？ did you run `cargo fmt`?

File: src/prover.rs [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    pub fn build_prove_error_response(&self, req: &ProveRequest, error_msg: &str) -> ProveResponse {
+        ProveResponse {
+            task_id: String::new(),
+            circuit_type: req.circuit_type,
+            circuit_version: req.circuit_version.clone(),
+            hard_fork_name: req.hard_fork_name.clone(),
+            status: TaskStatus::Failed,
+            created_at: 0.0,
```
Comment (Gigatron) [inline src/prover.rs:165]: why is it 0.0 for a datetime?

Comment (ErdongChen61) [reply src/prover.rs:165]: it is float type in `ProveResponse`, which is defined in scroll proving sdk

File: src/prover.rs (hunk 9 above)
Comment (ErdongChen61) [reply src/prover.rs:41]: good catch!

Comment (ErdongChen61) [reply src/prover.rs:41]: hmm, `cargo fmt` did not generate any changes

**Review by cyphersnake — APPROVED** [review]

**Review by Gigatron — APPROVED** [review]

**Pull request merged**